devlab/tests/generate_load.py

```diff
@@ -280,37 +299,92 @@ def get_parameters_for_vm_creating(_vm):
+                    search_opts={'all_tenants': 1})
+                spawning_vms = [vm.id for vm in all_vms
+                                if vm.status == 'BUILD']
+                if len(spawning_vms) >= VM_SPAWNING_LIMIT:
+                    time.sleep(1)
+                else:
+                    break
+            else:
+                raise RuntimeError(
+                    'VMs with ids {0} were in "BUILD" state more than {1} '
+                    'seconds'.format(spawning_vms, TIMEOUT))
+
         def create_vms(vm_list):
+            vm_ids = []
             for vm in vm_list:
-                _vm = self.check_vm_state(self.novaclient.servers.create(
-                    **get_parameters_for_vm_creating(vm)))
+                wait_for_vm_creating()
+                _vm = self.novaclient.servers.create(
+                    **get_parameters_for_vm_creating(vm))
+                vm_ids.append(_vm.id)
                 if not vm.get('fip'):
                     continue
+                wait_vm_nic_created(_vm.id)
                 fip = self.neutronclient.create_floatingip(
                     {"floatingip": {"floating_network_id": self.ext_net_id}})
                 _vm.add_floating_ip(fip['floatingip']['floating_ip_address'])
-
-        create_vms(self.config.vms)
+            return vm_ids
+        
+        vms = create_vms(self.config.vms)
         for tenant in self.config.tenants:
             if not tenant.get('vms'):
                 continue
             self.switch_user(user=self.username, password=self.password,
                              tenant=tenant['name'])
-            create_vms(tenant['vms'])
+            vms.extend(create_vms(tenant['vms']))
         self.switch_user(user=self.username, password=self.password,
                          tenant=self.tenant)
+        for i in range(TIMEOUT):
+            if vms:
+                for vm in vms[:]:
+                    if self.check_vm_state(vm):
+                        vms.remove(vm)
+                time.sleep(1)
+            else:
+                break
+        else:
+            raise RuntimeError(
+                'Vms with ids {0} have not become to active state'.format(vms))
 
     def create_vm_snapshots(self):
+        snp_ids = []
         for snapshot in self.config.snapshots:
             self.novaclient.servers.create_image(
                 server=self.get_vm_id(snapshot['server']),
                 image_name=snapshot['image_name'])
-            snp = self.glanceclient.images.get(self.get_image_id(snapshot['image_name']))
-            while snp.status != 'active':
-                time.sleep(2)
-                snp = self.glanceclient.images.get(self.get_image_id(snapshot['image_name']))
-                if snp.status == 'error':
-                    return None
+            snp = self.glanceclient.images.get(self.get_image_id(
+                snapshot['image_name']))
+            snp_ids.append(snp.id)
+        for i in range(TIMEOUT):
```

Comment:
I see `for i in range(TIMEOUT)` loop for the 4th time. Please consider refactoring, keep [DRY](https://en.wikipedia.org/wiki/Don't_repeat_yourself).